src/app/apps/page.tsx

```diff
@@ -1,43 +1,34 @@
-import Link from "next/link";
-import styles from "./apps.module.css";
+"use client"
+
+import {AppList} from "@/components/AppList";
+import {SessionProvider} from "next-auth/react";
+import {LoginButton} from "@/components/LoginButton";
+import {useEffect, useState} from "react";
 import {
-  getApps,
-  getCategories,
-  getDevices,
+    getAppsResponse,
+    getCategoriesResponse,
+    getDevicesResponse
 } from "@/badgehub-api-client/generated/swagger/public/public";
-import { Filter } from "@/components/Filter";
+import {getAppData} from "../actions";
 
 export interface SearchParams {
-  category: string;
-  device: string;
+    category: string;
+    device: string;
 }
 
-export default async function Listing({
-  searchParams,
-}: {
-  searchParams: Partial<SearchParams>;
+export default function Listing({ searchParams }: {
+    searchParams: Partial<SearchParams>;
 }) {
-  const [apps, categories, devices] = await Promise.all([
-    getApps(searchParams),
-    getCategories(),
-    getDevices(),
-  ]);
+    const [data, setData] = useState<[getAppsResponse, getCategoriesResponse, getDevicesResponse] | null>(null);
 
-  return (
-    <>
-      <Filter categories={categories.data} devices={devices.data} />
+    useEffect(() => {
+        getAppData(searchParams).then((data) => setData(data));
+    }, [searchParams]);
 
-      {apps.data.map((app) => (
-        <article className={styles.appCard} key={app.slug}>
-          <Link href={`/apps/${app.slug}`}>
-            <h2>{app.name}</h2>
-          </Link>
-          <Link href={`/categories/${app.category_slug}`}>
-            {app.category_slug}
-          </Link>
-          <p>{app.user_name}</p>
-        </article>
-      ))}
-    </>
-  );
+    return (
+        <SessionProvider>
+            <LoginButton/>
+            {data ? <AppList data={data}/> : <p>Loading new data...</p>}
```

Comment:
It would be great if we could SSR this

Reply:
To select the active link, we need `usePathname()`, which seems to be only available on the client. I would like to hear if you have a solution for this.

Reply:
There's a bit more nuance to that - `usePathname` is only available in *client components*, but client components do their first render on the server. By first rendering `<p>Loading new data...<p>` here, we make sure that user without JS enabled only ever see that. But by awaiting the data request directly rather than kicking it off in a use effect, the data is rendered into html on the server, and the rendered html hits the cache, which is faster. It can then rehydrate on the client, and we could use something like react query to manage client side caching and refreshing - and crucially, that's an example of progressive enhancement.

Reply:
Just want to let you know this is still on my to do list